### DISCUSSION TITLE
GraphQL WHERE clause on array field returns all items in array

### DISCUSSION THREAD

**Discussion by criticalmaas** [payloadcms/payload#428] [Feature Requests & Ideas]
# Bug Report

When using GraphQL with a collection that contains an array, when using the where clause on an array field which is of type `checkbox` or `select`, the result is that all items in the array are returned

## Expected Behavior

When using the where clause, only the items in the array that match the where condition should be returned. Or alternatively, can a where condition be supported on the nested array portion of the query?

## Steps to Reproduce

1. create a collection with an array
2. inside the array create either a check box (true/false) or select with two items. In the example I used "draft" and "published"
3. create two array elements in the document each with different values for the checkbox or select
4. create a GraphQL query to select items from the array given a where clause which equals the value of checkbox or select
5. notice that all items in the array are returned

## Detailed Description

collection fields:
``` typescript
import { CollectionConfig } from "payload/dist/collections/config/types"

const LocationsList: CollectionConfig = {
  slug: "locations-list",
  fields: [
    {
      name: "region",
      type: "relationship",
      relationTo: "regions",
      hasMany: false,
      localized: true,
    },
    {
      name: "locations",
      type: "array",
      label: "Locations within Region",
      labels: {
        singular: "Location",
        plural: "Locations",
      },
      fields: [
        {
          type: "row",
          fields: [
            {
              name: "label",
              type: "text",
              localized: true,
            },
            {
              name: "value",
              type: "text",
            },
          ],
        },
        {
          name: "status",
          type: "select",
          options: ["draft", "published"],
        },
      ],
    },
  ],
}

export default LocationsList
```

sample query
``` graphql
{
  LocationsLists(
    where: { locations__status: { equals: draft}}) {
    docs {
      region {
        label
      }
      locations {
        label
        status
      }
    }
  }
}
```

response:
``` json
"data": {
    "LocationsLists": {
      "docs": [
        {
          "region": {
            "label": "California"
          },
          "isActive": true,
          "locations": [
            {
              "label": "Burbank",
              "status": "draft"
            },
            {
              "label": "Glendale",
              "status": "published"
            }
          ]
        },
        {
          "region": {
            "label": "Florida"
          },
          "isActive": true,
          "locations": [
            {
              "label": "Orlando",
              "status": "draft"
            }
          ]
        }
      ]
    }
  }
}

```

payload v0.13.3
browser: chrome latest

**Comment by jmikrut** [comment]
Hey @criticalmaas — thank you for bringing this up. I think this is working as expected right now, because the first document in your returned results _**does**_ have a `status` field equal to `draft`. 

Although:

> Or alternatively, can a where condition be supported on the nested array portion of the query?

This is totally possible. It would be a GraphQL-only feature, of course, but it's certainly doable. 

The bulk of the work would be to convert our database query syntax into something to be used in GraphQL resolvers on POJOs, which could end up being useful elsewhere as well.

If you're interested in pursuing this feature request, I suggest we close this issue and take the conversation to a new Feature Request Discussion where we can talk through the approach in further depth. 

This new feature could also be applied to the `blocks` field type and I think it would be important to get the query syntax easily abstracted to work against POJOs as well as MongoDB queries.

What do you think?

**Comment by freakyfriday** [comment]
Hi @jmikrut,

Firstly thanks for putting together such an awesome headless CMS framework! I was having a think about a solution to this problem and thought perhaps we could look at using `aggregation` in Mongoose with a `$filter` on an array property which requires only certain results to be returned.

Below is an example of what the query might look like:

<img width="749" alt="Screen Shot 2022-06-17 at 9 53 15 am" src="https://user-images.githubusercontent.com/8381608/174196409-050153fa-0b53-4c53-bf95-1487e416958a.png">

1. Each array property is queryable eg `services(where: $where_services)` which has a query type eg `User_Services_where`

2. If the array property has a query then we implement a `$filter` command in Mongoose using `aggregation`. Because the shape of the document should be the same we can cast Mongoose plain object back into a mongoose doc. See example - https://stackoverflow.com/a/36371665/3043981

What do you think?
